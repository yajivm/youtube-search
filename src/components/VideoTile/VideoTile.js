import React from 'react';
import PropTypes from 'prop-types';
import { getPublishedDate } from '../../helpers/utils';

import './VideoTile.style.css';

const VideoTile = ({ videoData, onSelectVideoTile }) => {
    const selectVideo = () => {
        onSelectVideoTile(videoData);
    };

    const getVideoChannelDetails = () => {
        if (videoData && Object.keys(videoData?.snippet).length > 0) {
            return (
                <ul className="details-list">
                    <li data-testid="channel-title">{videoData?.snippet?.channelTitle}</li>
                    <li data-testid="published-date">{getPublishedDate(videoData?.snippet?.publishedAt)}</li>
                </ul>
            )
        }
    };

    return (
        <div
            className="video"
            onClick={() => selectVideo()}
            data-testid={`video${videoData?.id?.videoId}`}
        >
            <div className="video-thumbnail">
                <img src={videoData?.snippet?.thumbnails?.high?.url} alt="videoTile" title={videoData?.snippet?.title} />
            </div>
            <div className="video-details">
                <p className="video-title" data-testid="video-title">{videoData?.snippet?.title}</p>
                {getVideoChannelDetails()}
            </div>
        </div>
    );
};

VideoTile.propTypes = {
    videoData: PropTypes.shape({
        id: PropTypes.shape({
            videoId: PropTypes.string,
        }),
        snippet: PropTypes.shape(),
    }).isRequired,
    onSelectVideoTile: PropTypes.func.isRequired,
};

export default VideoTile;
