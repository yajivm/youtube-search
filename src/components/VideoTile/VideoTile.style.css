.video {
    display: flex;
    flex-direction: row;
    border-radius: 2px;
    margin: 0 20px 20px;
    cursor: pointer;
}

.video-thumbnail {
    display: flex;
    width: 35%;
    flex-direction: column;
}

.video-thumbnail img {
    width: 135px;
}

.video-title {
    padding: 0 10px 10px 20px;
    margin: 0;
    font-weight: 600;
    font-size: 14px;
}

.video .details-list {
    list-style: none;
    font-size: 12px;
    margin: 0;
    padding: 0 0 0 20px;
}

.video .details-list > li {
  display: inline;
}

.video .details-list > li:first-child::after {
    content: "â€¢";
    margin: 0 4px;
}