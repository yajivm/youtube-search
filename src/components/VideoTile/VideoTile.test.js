import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import VideoTile from './VideoTile';

const mockGetPublishedData = jest.fn();
jest.mock('../../helpers/utils', () => {
    return {
        getPublishedDate: args => mockGetPublishedData(args),
    }
});

let mockVideoData = {
    id: {
        videoId: 'aTestId',
    },
    snippet: {
        publishedAt: "2021-11-12T22:18:07Z",
        title: "aTestTitle",
        channelId: "aTestChannelId",
        thumbnails: {
            high: {
                url: "aTestUrl",
            }
        },
    }
};
const mockOnSelectedVideoTitle = jest.fn();

test('should render video tile component', () => {
    const { getByTestId } = render(<VideoTile videoData={mockVideoData} onSelectVideoTile={mockOnSelectedVideoTitle} />);
    
    const videoTileWrapper = getByTestId('videoaTestId');

    expect(videoTileWrapper).toBeDefined();
    expect(videoTileWrapper).toMatchSnapshot();
});

test('should render video title when video title prop is valid', () => {
    mockVideoData.snippet.title = 'aTestTitle';

    const { getByTestId } = render(<VideoTile videoData={mockVideoData} onSelectVideoTile={mockOnSelectedVideoTitle} />);
    
    expect(getByTestId('video-title')).toBeDefined();
    expect(getByTestId('video-title').textContent).toStrictEqual('aTestTitle');
});

test('should render channel details and date when publishedAt and channelTitle props are valid', () => {
    jest.resetAllMocks();
    mockVideoData.snippet.channelTitle = 'aChannelTestTitle';
    mockVideoData.snippet.publishedAt = '2021-11-12T22:18:07Z';

    const { getByTestId } = render(<VideoTile videoData={mockVideoData} onSelectVideoTile={mockOnSelectedVideoTitle} />);
    
    expect(getByTestId('channel-title')).toBeDefined();
    expect(getByTestId('channel-title').textContent).toStrictEqual('aChannelTestTitle');
    expect(mockGetPublishedData).toHaveBeenCalledTimes(1);
    expect(mockGetPublishedData).toHaveBeenCalledWith('2021-11-12T22:18:07Z');
    expect(getByTestId('published-date')).toBeDefined();
});

test('should call call back when click video tile', () => {
    jest.resetAllMocks();
    mockVideoData.snippet.channelTitle = 'aChannelTestTitle';
    mockVideoData.snippet.publishedAt = '2021-11-12T22:18:07Z';

    const { getByTestId } = render(<VideoTile videoData={mockVideoData} onSelectVideoTile={mockOnSelectedVideoTitle} />);
    
    const videoTileWrapper = getByTestId('videoaTestId');

    expect(videoTileWrapper).toBeDefined();

    fireEvent.click(videoTileWrapper);

    expect(mockOnSelectedVideoTitle).toHaveBeenCalledTimes(1);
    expect(mockOnSelectedVideoTitle).toHaveBeenCalledWith(mockVideoData);
});