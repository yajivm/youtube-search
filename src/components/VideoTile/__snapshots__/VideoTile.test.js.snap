// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render video tile component 1`] = `
<div
  class="video"
  data-testid="videoaTestId"
>
  <div
    class="video-thumbnail"
  >
    <img
      alt="videoTile"
      src="aTestUrl"
      title="aTestTitle"
    />
  </div>
  <div
    class="video-details"
  >
    <p
      class="video-title"
      data-testid="video-title"
    >
      aTestTitle
    </p>
    <ul
      class="details-list"
    >
      <li
        data-testid="channel-title"
      />
      <li
        data-testid="published-date"
      />
    </ul>
  </div>
</div>
`;
